webapi/setaltsig.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package webapi
+
+import (
+	"encoding/base64"
+	"time"
+
+	"github.com/decred/dcrd/chaincfg/v3"
+	dcrdtypes "github.com/decred/dcrd/rpc/jsonrpc/types/v3"
+	"github.com/decred/dcrd/txscript/v4/stdaddr"
+	"github.com/decred/vspd/database"
+	"github.com/decred/vspd/rpc"
+	"github.com/gin-gonic/gin"
+	"github.com/gin-gonic/gin/binding"
+)
+
+// Ensure that Node is satisfied by *rpc.DcrdRPC.
+var _ Node = (*rpc.DcrdRPC)(nil)
+
+// Node is satisfied by *rpc.DcrdRPC and retrieves data from the blockchain.
+type Node interface {
+	CanTicketVote(rawTx *dcrdtypes.TxRawResult, ticketHash string, netParams *chaincfg.Params) (bool, error)
+	GetRawTransaction(txHash string) (*dcrdtypes.TxRawResult, error)
+}
+
+// setAltSig is the handler for "POST /api/v3/setaltsig".
+func setAltSig(c *gin.Context) {
+
+	const funcName = "setAltSig"
+
+	// Get values which have been added to context by middleware.
+	sigAddress := c.MustGet("SigningAddress").(string)
+	dcrdClient := c.MustGet("DcrdClient").(Node)
+	reqBytes := c.MustGet("RequestBytes").([]byte)
+
+	if cfg.VspClosed {
+		sendError(errVspClosed, c)
+		return
+	}
+
+	var request SetAltSigRequest
+	if err := binding.JSON.BindBody(reqBytes, &request); err != nil {
+		log.Tracef("%s: Bad request (clientIP=%s): %v", funcName, c.ClientIP(), err)
+		sendErrorWithMsg(err.Error(), errBadRequest, c)
+		return
+	}
+
+	altPubAddr, ticketHash := request.AltPubAddress, request.TicketHash
+
+	currentAltSigAddr, err := db.AltSigAddr(ticketHash)
+	if err != nil {
+		log.Errorf("%s: db.AltSigExists (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+	if currentAltSigAddr != "" {
+		msg := "alternate signature address already exists"
+		log.Errorf("%s: %s (ticketHash=%s)", funcName, msg, ticketHash)
+		sendErrorWithMsg(msg, errBadRequest, c)
+		return
+
+	}
+
+	// Disallow empty addresses.
+	if altPubAddr == "" {
```

Comment:
This will be unnecessary if we add json annotations to make it required.

Reply:
Ok.

Reply:
0175871